### ISSUE TITLE
Some workarounds for DAdaption loss going NaN

### ISSUE THREAD

**Issue by thojmr** [kohya-ss/sd-scripts#592]
I've been having an issue where loss goes to NaN when using DAdaptation, but after days of debugging I came up with a few things that can cause this.  (DAdaptation is so fragile...)

This applies to LoRA and LoCon model training:
- I have to include betas as an optimizer arg for DAdaptAdam. `--optimizer_args "decouple=True" "weight_decay=1.0" "betas=0.9,0.99"` Not sure why.  Otherwise loss goes NaN. (weight_decay can vary 1.0-0.01)
- Don't include a `--vae`, or at least not the Av3 VAE, else loss goes NaN
- Don't accidentally include `--max_token_length` if you are not training with captions.  Loss will go NaN if set above default.
  - edit: even when training with captions I get NaN after a hundreds steps when setting this value above default :(
- Alpha must match Dim.  Leaving Alpha=1, loss will go NaN, though a little slower than the rest

I'm still tinkering but ill update with more details if I find any other bad arguments.
Hope it helps, DAdaptation is promising, but a pain to use so far.

related: https://github.com/kohya-ss/sd-scripts/issues/570

**Comment by sdbds** [comment]
can you provide you config about steps?
image number、repeat、epoch and so on...

**Comment by thojmr** [comment]
Sure. This was a small character dataset:

- 6 images
- default repeat

```
steps=1000  # usually NaN by 100th step
learning_rate=1
train_batch_size=2
gradient_accumulation_steps=1
mixed_precision="bf16"
save_every_n_epochs=10
max_resolution=768
max_resolution_bucket=768
LR_SCHEDULER="constant" 
network_dim=8
alpha=8
tokens=225
tag_dropout=0.15
dropout=0.3

accelerate launch --num_cpu_threads_per_process $num_cpu_threads_per_process train_network.py \
    --pretrained_model_name_or_path="${pretrained_model_name_or_path}" \
    --train_data_dir="../images/${raw_img_folder}" \
    --output_dir="./output/${short_name}" \
    --resolution=$max_resolution \
    --train_batch_size=$train_batch_size \
    --gradient_checkpointing \
    --gradient_accumulation_steps=$gradient_accumulation_steps \
    --learning_rate=${learning_rate} \
    --lr_scheduler="${LR_SCHEDULER}" \
    --max_train_steps=$steps \
    --xformers \
    --mixed_precision=$mixed_precision \
    --save_precision=$save_precision \
    --save_every_n_epochs=$save_every_n_epochs \
    --save_model_as="safetensors" \
    --clip_skip=2 \
    --seed=92386 \
    --flip_aug \
    --network_module=networks.lora \
    --enable_bucket \
    --no_metadata \
    --max_bucket_reso=$max_resolution_bucket \
    --shuffle_caption \
    --logging_dir=logs \
    --log_prefix="${short_name}_lr${learning_rate}_" \
    --output_name="${FULL_NAME}" \
    --network_dim=$network_dim \
    --network_alpha=$alpha \
    --caption_extension ".txt" \
    --caption_tag_dropout_rate $tag_dropout \
    --max_data_loader_n_workers 4 \
    --lr_scheduler_num_cycles 33 \
    --network_args "conv_dim=${network_dim}" "conv_alpha=${alpha}" \
    --network_train_unet_only \
    --optimizer_type "DAdaptAdam" \
    --optimizer_args "decouple=True" "weight_decay=1.0" "betas=0.9,0.99" \
    --sample_every_n_epochs 10
```

**Comment by sdbds** [comment]
I think the issue with warmup might be due to the use of constant, as dadapt tends to cause overfitting quickly. 
I recommend using the latest optimizer Prodigy, which is very effective. 
I will soon publish a related tutorial article on Civitai.

**Comment by thojmr** [comment]
Had some success with DAdaptation + cosine_with_restarts.  I just had to ignore all the others here mentioning that you should use `constant` with DAdaptation ¯\\\_(ツ)\_/¯

But I agree that DAdaptation learns much faster than ADAfactor.  You can see DAdaptation reach peak LR in a fraction of the time it takes ADAFactor to do the same.  ADAFactor still seems more hands-off and stable overall, but ill have to compare results to tell which I want to stick with.
